apps/deploy-web/src/services/provider-proxy/provider-proxy.service.ts

```diff
@@ -76,6 +87,202 @@ export class ProviderProxyService {
+      options.start?.(state, ws);
+    });
+  }
+
+  async downloadLogs(input: {
+    providerBaseUrl: string;
+    providerAddress: string;
+    providerCredentials: ProviderCredentials;
+    dseq: string;
+    gseq: number;
+    oseq: number;
+    type: "logs" | "events";
+    chainNetwork: string;
+    signal?: AbortSignal;
+  }): Promise<DownloadMessagesResult> {
+    const baseUrl = `${input.providerBaseUrl}/lease/${input.dseq}/${input.gseq}/${input.oseq}`;
+    const url = input.type === "logs" ? `${baseUrl}/logs?follow=false&tail=10000000` : `${baseUrl}/kubeevents?follow=false&tail=10000000`;
+
+    let logFileContent = "";
+    let lastMessageTimestamp = Date.now();
+    const result = await this.downloadMessages(url, {
+      signal: input.signal,
+      providerAddress: input.providerAddress,
+      chainNetwork: input.chainNetwork,
+      providerCredentials: input.providerCredentials,
+      start: (state, ws) => {
+        setTimeout(function tick() {
+          const elapsed = Date.now() - lastMessageTimestamp;
+
+          if (elapsed > 3_000) {
+            state.isFinished = true;
+            if (ws.readyState !== WebSocket.CLOSED && ws.readyState !== WebSocket.CLOSING) {
+              ws.close();
+            }
+          } else {
+            setTimeout(tick, 1_000);
+          }
+        }, 1_000);
+      },
+      onMessage: event => {
+        const data = JSON.parse(event.data);
+        if (data.closed || !data.message) return;
+
+        const parsedLog = JSON.parse(data.message);
+
+        let service: string;
+        let message: string;
+        if (input.type === "logs") {
+          service = parsedLog?.name ? parsedLog?.name.split("-")[0] : "";
+          message = `[${service}]: ${parsedLog.message}`;
+        } else {
+          service = parsedLog.object?.name ? parsedLog.object?.name.split("-")[0] : "";
+          message = `[${service}]: [${parsedLog.type}] [${parsedLog.reason}] [${parsedLog.object?.kind}] ${parsedLog.note}`;
+        }
+
+        logFileContent += message + "\n";
+        lastMessageTimestamp = Date.now();
+      }
+    });
+
+    if (result.ok) {
+      const fileName = `${input.dseq}-${input.gseq}-${input.oseq}-${input.type}-${new Date().toISOString().substring(0, 10)}.txt`;
+      this.saveFile(new Blob([logFileContent], { type: "text/plain" }), fileName);
+    }
+    return result;
+  }
+
+  async downloadFileFromShell(input: {
+    providerBaseUrl: string;
+    providerAddress: string;
+    providerCredentials: ProviderCredentials;
+    dseq: string;
+    gseq: number;
+    oseq: number;
+    chainNetwork: string;
+    signal?: AbortSignal;
+    service: string;
+    filePath: string;
+  }): Promise<DownloadMessagesResult> {
+    const printCommand = "cat"; // print command on Windows is "type"
+    const command = `${printCommand} ${input.filePath}`;
+    const url = `${input.providerBaseUrl}/lease/${input.dseq}/${input.gseq}/${input.oseq}/shell?stdin=0&tty=0&podIndex=0${command
```

Comment:
comment(non-blocking): I'm sure if we have to now, but after the new upgrade we might have to add `bseq` to this, but I don't think it's implemented and we have to do anything there, just an observation.